pkg/cmd/cli/cmd/rollback.go

```diff
@@ -0,0 +1,122 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	kubectl "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl"
+	kcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd"
+
+	describe "github.com/openshift/origin/pkg/cmd/cli/describe"
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+)
+
+const rollbackLongDesc = `
+Revert part of an application back to a previous deployment.
+
+When you run this command your deployment configuration will be updated to match
+the provided deployment. By default only the pod and container configuration
+will be changed and scaling or trigger settings will be left as-is. Note that
+environment variables and volumes are included in rollbacks, so if you've
+recently updated security credentials in your environment your previous
+deployment may not have the correct values.
+
+If you would like to review the outcome of the rollback, pass '--dry-run' to print
+a human-readable representation of the updated deployment configuration instead of
+executing the rollback. This is useful if you're not quite sure what the outcome
+will be.
+
+Examples:
+  Perform a rollback:
+
+  $ %[1]s %[2]s deployment-1
+
+  See what the rollback will look like, but don't perform the rollback:
+
+  $ %[1]s %[2]s deployment-1 --dry-run
+
+  Perform the rollback manually by piping the JSON of the new config back to %[1]s:
+
+  $ %[1]s %[2]s deployment-1 --output=json | %[1]s update deploymentConfigs deployment -f -
+`
+
+func NewCmdRollback(parentName string, name string, f *Factory, out io.Writer) *cobra.Command {
+	rollback := &deployapi.DeploymentConfigRollback{
+		Spec: deployapi.DeploymentConfigRollbackSpec{
+			IncludeTemplate: true,
+		},
+	}
+
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s <from-deployment>", name),
+		Short: "Revert part of an application back to a previous deployment.",
+		Long:  fmt.Sprintf(rollbackLongDesc, parentName, name),
+		Run: func(cmd *cobra.Command, args []string) {
+			if len(args) == 0 || len(args[0]) == 0 {
+				usageError(cmd, "A deployment name is required.")
+			}
+
+			rollback.Spec.From.Name = args[0]
+
+			outputFormat := kcmd.GetFlagString(cmd, "output")
+			outputTemplate := kcmd.GetFlagString(cmd, "template")
+			dryRun := kcmd.GetFlagBool(cmd, "dry-run")
+
+			osClient, _, err := f.Clients(cmd)
+			checkErr(err)
+
+			namespace := getOriginNamespace(cmd)
+
+			// Generate the rollback config
+			newConfig, err := osClient.DeploymentConfigs(namespace).Rollback(rollback)
+			checkErr(err)
+
+			// If dry-run is specified, describe the rollback and exit
+			if dryRun {
+				describer := &describe.DeploymentConfigDescriber{
+					Client: &rollbackDeploymentDescriberClient{
```

Comment:
You can easily make Describer for DeploymentConfig ignore namespace, name.

----- Original Message -----
> > +			outputTemplate := kcmd.GetFlagString(cmd, "template")
> > +			dryRun := kcmd.GetFlagBool(cmd, "dry-run")
> > +
> > +			osClient, _, err := f.Clients(cmd)
> > +			checkErr(err)
> > +
> > +			namespace := getOriginNamespace(cmd)
> > +
> > +			// Generate the rollback config
> > +			newConfig, err :=
> > osClient.DeploymentConfigs(namespace).Rollback(rollback)
> > +			checkErr(err)
> > +
> > +			// If dry-run is specified, describe the rollback and exit
> > +			if dryRun {
> > +				describer := &describe.DeploymentConfigDescriber{
> > +					Client: &rollbackDeploymentDescriberClient{
> 
> Because the upstream describer interface is this:
> 
> ```go
> type Describer interface {
> 	Describe(namespace, name string) (output string, err error)
> }
> ```
> 
> Every other Describer impl takes a client. If you make a Describer via
> constructor that takes a DeploymenConfig, you still call Describe(string,
> string), which makes no sense. I'm doing what I can without changing the
> interface or general patterns. The code I have works, is relatively
> consistent with the other code, and  produces the right output.
> 
> Why don't we deal with Describer refactors in general separately? They all
> need work.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/786/files#r23871300